Composer/packages/server/src/models/connector/csharpBotConnector.ts

```diff
@@ -1,92 +1,205 @@
+        currentDebugger('bot runtime (%d): %s', child.pid, data);
+        resolve(child.pid);
+      });
+
+    child.stderr &&
+      child.stderr.on('data', (err: any) => {
+        erroutput += err.toString();
+      });
+
+    child.on('exit', code => {
+      currentDebugger('exit %d', code);
+      handler();
+      if (code !== 0) {
+        currentDebugger('exit %d: %s', code, erroutput);
+        reject(erroutput);
+      }
+    });
+
+    child.on('message', msg => {
+      currentDebugger('bot runtime received: %s', msg);
+    });
+  };
+
+  private getBotPath = () => {
     const currentProject = BotProjectService.getCurrentBotProject();
     if (currentProject === undefined) {
       throw new Error('no project is opened, nothing to sync');
     }
-    const dir = Path.join(currentProject.dataDir);
-    const luisConfig = currentProject.luPublisher.getLuisConfig();
-    await this.archiveDirectory(dir, './tmp.zip');
-    const content = fs.readFileSync('./tmp.zip');
+    return Path.join(currentProject.dir);
+  };
 
-    const form = new FormData();
-    form.append('file', content, 'bot.zip');
+  private start = async (dir: string, config: DialogSetting): Promise<string> => {
+    return new Promise((resolve, reject) => {
+      const runtime = spawn(
+        'dotnet',
+        [
+          'bin/Debug/netcoreapp2.1/BotProject.dll',
+          `--urls`,
+          this.endpoint,
+          ...this.getConnectorConfig(config),
+          '--environment',
+          'development',
+        ],
+        {
+          detached: true,
+          cwd: dir,
+          stdio: ['ignore', 'pipe', 'pipe'],
+        }
+      );
+      // extend runtime debugger
+      runtimeDebugs[runtime.pid] = log.extend(`port ${runtime.pid}`);
+      runtimeDebugs[runtime.pid]('bot runtime started. pid: %d', runtime.pid);
+      CSharpBotConnector.botRuntimes[runtime.pid] = runtime;
+      this.addListeners(runtime, this.stop, resolve, reject);
+    });
+  };
 
-    if (luisConfig && luisConfig.authoringKey !== null && !currentProject.checkLuisPublished()) {
-      throw new Error('Please publish your Luis models');
+  private checkPortUsable = async (port: string | number): Promise<string> => {
+    if (process.platform === 'win32') {
+      return new Promise((resolve, reject) => {
+        const netstat = spawn('netstat', ['-ano']);
+        const findstr = spawn('findstr', [`${port}`], {
+          stdio: [netstat.stdout, 'pipe', process.stderr],
+        });
+        let output = '';
+        findstr.stderr &&
+          findstr.stderr.on('data', function(err) {
+            reject(err.toString());
+          });
+        findstr.stdout && findstr.stdout.on('data', data => (output += data));
+        findstr.on('exit', function() {
+          resolve(output);
+        });
+      });
+    } else {
+      return new Promise((resolve, reject) => {
+        const lsof = spawn('sudo', ['lsof', '-i', `:${port}`]);
```

Comment:
This is going to be a no go as well. 

Let's just use https://www.npmjs.com/package/get-port and assign a port instead of erroring when a port is already in use.

Reply:
remove this, and use getPort instead.